### PR TITLE
Stop copying truss server code over for "custom server"

### PR DESCRIPTION
**Pull request by tianshuc0731**
## :rocket: What

We caught some bugs in trussless custom server in real life when customers try to start their server using `python main.py` as start command
https://basetenlabs.slack.com/archives/C037X7HFAE4/p1728672357619419?thread_ts=1728574438.184139&cid=C037X7HFAE4

## :computer: How
- stop copying over truss code to container in "trussless" mode
- stop overriding WORKDIR in "trussless" mode
- remove torch dependency from trussless test case for faster image building and test running
- change trussless test case to use `python main.py` to start fastapi server, so we can catch similar type of bugs in the future
- some trussless files renaming
- [non-related patch] make both `docker_server.liveness_endpoint` and `docker_server.readiness_endpoint` optional, also add assertion to check all the other required fields are provided 

## :microscope: Testing
Passed integration test
https://github.com/basetenlabs/truss/actions/runs/12126088093

### CODE REVIEW

**Review by squidarth — APPROVED** [review]
lgtm. 

Might also be worth trying out dhruv's config before merging

**Comment by tianshuc0731** [comment]
> lgtm.
> 
> Might also be worth trying out dhruv's config before merging

@dsingal0 can you try context image "v0.9.44-rc" to verify it can resolve the issue?

**Comment by squidarth** [comment]
@tianshuc0731 I also think that we might need to merge https://github.com/basetenlabs/truss/pull/1187 to properly test dhruv's changes. it lgtm

**Comment by tianshuc0731** [comment]
> @tianshuc0731 I also think that we might need to merge #1187 to properly test dhruv's changes. it lgtm

Merged the change, @dsingal0  please test this new image "v0.9.44-rc2"

File: pyproject.toml
Comment (tianshuc0731) [inline pyproject.toml:3]: Should I make change in this pr? @squidarth

File: truss/truss_handle.py
Comment (tianshuc0731) [inline truss/truss_handle.py:1070]: @squidarth The motivation is to make trussless integration test fail faster when the container successfully built but failed to start the server, currently we need to wait 120 seconds to see that failure. I want to make sure it's not breaking anything.

File: pyproject.toml
Comment (squidarth) [reply pyproject.toml:3]: you don't need to -- also, it's more typical to say "0.9.44rc2". Also increment the patch version, this should technically be "0.9.45rc2", since 0.9.44 is already released

File: truss/truss_handle.py
Comment (squidarth) [reply truss/truss_handle.py:1070]: is this used anywhere outside of the tests?

Comment (squidarth) [reply truss/truss_handle.py:1070]: yes, it does look like it is. Could you change this to have the "stop_after_delay" passed as a parameter?

You can switch to using this instead: https://tenacity.readthedocs.io/en/latest/index.html#changing-arguments-at-run-time

File: truss/truss_handle.py
Comment (squidarth) [inline truss/truss_handle.py]: instead of duplicating the decorator, can you do something like modify `_wait_for_model_server` to be this:

```
def _wait_for_model_server(url: str, wait_seconds: int = 120) -> Response:
    for attempt in Retrying(...):
        with attempt:
            return requests.get(...)
```

File: truss/templates/server.Dockerfile.jinja
Comment (squidarth) [inline truss/templates/server.Dockerfile.jinja:111]: one non-blocking comment, since this resulted in an actual bug/issue, would be awesome if we could also have an integration test for this, don't know how easy that would be to put together

Comment (tianshuc0731) [reply truss/templates/server.Dockerfile.jinja:111]: No easy way on top of my mind, maybe what we want is to have a different testing docker image which installs supervisord in a different path? Any better ideas?

Comment (squidarth) [reply truss/templates/server.Dockerfile.jinja:111]: yeah maybe something like that. no need to block on this then

File: truss/base/truss_config.py
Comment (tianshuc0731) [inline truss/base/truss_config.py:427]: Can you help me confirm these default values can work @squidarth @bolasim so we can make both `readiness_endpoint` and `liveness_endpoint` optional in the config.yaml.

Comment (bolasim) [reply truss/base/truss_config.py:427]: the endpoints are enumerated here: https://github.com/basetenlabs/truss/blob/bd609921800c0bed324358e38d70948880a035d8/truss/templates/server/truss_server.py#L285-L290

Comment (tianshuc0731) [reply truss/base/truss_config.py:427]: I see then I will just skip forward logic in proxy config when they are not configured

**Pull request merged**
